pkg/cloudprovider/providers/photon/photon.go

```diff
@@ -0,0 +1,473 @@
+type PCCloud struct {
+	cfg *PCConfig
+	// InstanceID of the server where this PCCloud object is instantiated.
+	localInstanceID string
+	// local $HOSTNAME
+	localHostname string
+	// hostname from K8S, could be overridden
+	localK8sHostname string
+	// Photon Project ID
+	projID string
+	//	instance      *Instances
+	cloudprovider.Zone
+}
+
+type PCConfig struct {
+	Global struct {
+		CloudTarget       string `gcfg:"target"`
+		IgnoreCertificate bool   `gcfg:"ignoreCertificate"`
+		Tenant            string `gcfg:"tenant"`
+		Project           string `gcfg:"project"`
+		OverrideIP        bool   `gcfg:"overrideIP"`
+	}
+}
+
+// Disks is interface for manipulation with PhotonController Persistent Disks.
+type Disks interface {
+	// AttachDisk attaches given disk to given node. Current node
+	// is used when nodeName is empty string.
+	AttachDisk(pdID string, nodeName k8stypes.NodeName) error
+
+	// DetachDisk detaches given disk to given node. Current node
+	// is used when nodeName is empty string.
+	DetachDisk(pdID string, nodeName k8stypes.NodeName) error
+
+	// DiskIsAttached checks if a disk is attached to the given node.
+	DiskIsAttached(pdID string, nodeName k8stypes.NodeName) (bool, error)
+
+	// CreateDisk creates a new PD with given properties.
+	CreateDisk(volumeOptions *VolumeOptions) (pdID string, err error)
+
+	// DeleteDisk deletes PD.
+	DeleteDisk(pdID string) error
+}
+
+// VolumeOptions specifies capacity, tags, name and flavorID for a volume.
+type VolumeOptions struct {
+	CapacityGB int
+	Tags       map[string]string
+	Name       string
+	Flavor     string
+}
+
+func readConfig(config io.Reader) (PCConfig, error) {
+	if config == nil {
+		err := fmt.Errorf("no Photon Cloud Provider config file given")
+		return PCConfig{}, err
+	}
+
+	var cfg PCConfig
+	err := gcfg.ReadInto(&cfg, config)
+	return cfg, err
+}
+
+func init() {
+	cloudprovider.RegisterCloudProvider(ProviderName, func(config io.Reader) (cloudprovider.Interface, error) {
+		cfg, err := readConfig(config)
+		if err != nil {
+			return nil, err
+		}
+		return newPCCloud(cfg)
+	})
+}
+
+func getVMIDbyNodename(project string, nodeName string) (string, error) {
+	vmList, err := Esxclient.Projects.GetVMs(project, nil)
+	if err != nil {
+		glog.V(2).Info("Photon Cloud Provider: Failed to GetVMs from project %s with nodeName %s", project, nodeName)
+		return "", err
+	}
+
+	for _, vm := range vmList.Items {
+		if vm.Name == nodeName {
+			// TODO: check vm state
+			return vm.ID, nil
+		}
+	}
+
+	return "", fmt.Errorf("Photon Cloud Provider: no matching VM is found with name %s", nodeName)
+}
+
+func getVMIDbyIP(project string, IPAddress string) (string, error) {
+	vmList, err := Esxclient.Projects.GetVMs(project, nil)
+	if err != nil {
+		glog.V(2).Info("Photon Cloud Provider: Failed to GetVMs from project %s, IP %s", project, IPAddress)
+		return "", err
+	}
+
+	for _, vm := range vmList.Items {
```

Comment:
No we need to get all the VMs in this project. The only knowledge we have for VM is the IP address. So we need to get all the VMs and compare the IP addresses with each VM, in order to find the right VM that we are looking for.